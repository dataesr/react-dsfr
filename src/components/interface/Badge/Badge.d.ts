import * as React from 'react';

export type BadgeColorFamily = any | '';

export type BadgeType = 'success' | 'error' | 'warning' | 'new' | 'info';

export interface BadgeProps {
  text: string;
  icon?: string;
  hasIcon?: boolean;
  isSmall?: boolean;
  colorFamily?: BadgeColorFamily;
  type?: BadgeType;
  className?: string;
}

declare const Badge: React.FC<BadgeProps>;

export default Badge;
