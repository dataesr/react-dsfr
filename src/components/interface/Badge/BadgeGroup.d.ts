import * as React from 'react';

export type BadgeGroupClassName = string | Object | any[];

export type BadgeGroupChildren = React.ReactNode[] | React.ReactNode;

export interface BadgeGroupProps {
  className?: BadgeGroupClassName;
  children?: BadgeGroupChildren;
}

declare const BadgeGroup: React.FC<BadgeGroupProps>;

export default BadgeGroup;
