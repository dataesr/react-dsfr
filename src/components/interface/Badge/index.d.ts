import type Badge from "./Badge";
import type BadgeGroup from "./BadgeGroup";

export type { Badge, BadgeGroup };
