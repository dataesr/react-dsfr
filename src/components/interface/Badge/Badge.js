import React from 'react';

import PropTypes from 'prop-types';
import classNames from 'classnames';
import Icon from '../../foundation/icon/index';

import dataAttributes from '../../../utils/data-attributes';
import '@gouvfr/dsfr/dist/component/badge/badge.css';
import colorFamilies from '../../../utils/variables';

/**
 *
 * @visibleName Badge
 */
const Badge = ({
  className,
  type,
  colorFamily,
  isSmall,
  text,
  icon,
  hasIcon,
  ...remainingProps
}) => {
  const _className = classNames(
    'fr-badge',
    {
      [`fr-badge--${type}`]: type,
      [`fr-badge--${colorFamily}`]: colorFamily,
      'fr-badge--no-icon': !hasIcon || !type,
      'fr-badge--sm': isSmall,
    },
    className,
  );

  const iconsType = {
    success: 'ri-checkbox-circle-fill',
    error: 'ri-settings-6-fill',
    info: 'ri-information-fill',
    warning: 'ri-error-warning-fill',
    new: 'ri-flashlight-fill',
  };

  const renderContent = () => {
    let content = text;

    if (hasIcon || icon) {
      content = (
        <Icon
          verticalAlign="sub"
          name={icon || iconsType[type]}
          size="lg"
          className="ds-fr-badge-icon"
        >
          {text}
        </Icon>
      );
    }
    return content;
  };

  return (
    <p
      className={_className}
      {...dataAttributes.getAll(remainingProps)}
    >
      {renderContent()}
    </p>
  );
};

Badge.defaultProps = {
  isSmall: false,
  hasIcon: false,
  className: '',
  colorFamily: '',
  type: 'new',
  icon: '',
};

Badge.propTypes = {
  text: PropTypes.string.isRequired,
  icon: PropTypes.string,
  hasIcon: PropTypes.bool,
  isSmall: PropTypes.bool,
  colorFamily: PropTypes.oneOf([...colorFamilies, '']),
  type: PropTypes.oneOf(['success', 'error', 'warning', 'new', 'info']),
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array,
  ]),
};

export default Badge;
