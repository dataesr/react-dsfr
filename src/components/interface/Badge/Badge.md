Le composant badge permet de mettre en avant une information de type “statut” ou “état” sur un élément du site.
Les icônes sont disponibles uniquement pour les badges systèmes ('success', 'error', 'warning', 'new', 'info').

### Success

```jsx
<Badge text="Success" type="success" colorFamily="brown-caramel"/>
```

### Erreur
A utiliser quand il y a plusieurs erreurs dans un formulaire, ou des erreurs bloquantes à remonter pour l’utilisateur.

```jsx
<Badge text="error" type="error"/>
```

### Small

```jsx
<Badge small text="Information" icon={false}/>
```
