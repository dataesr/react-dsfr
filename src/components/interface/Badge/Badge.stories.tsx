import { ComponentStory } from '@storybook/react';

import Badge from './Badge';

export default {
  title: 'Badge',
  component: Badge,
};

const Template: ComponentStory<typeof Badge> = (args) => (
  <Badge {...args} />
);

export const Succes = Template.bind({});
Succes.args = { text: 'Success', type: 'success', hasIcon: true };
Succes.storyName = 'Succes';

export const Nouveau = Template.bind({});
Nouveau.args = { text: 'Nouveau', type: 'new', hasIcon: true };
Nouveau.storyName = 'Nouveau';

export const Information = Template.bind({});
Information.args = { text: 'Info', type: 'info', hasIcon: true };
Information.storyName = 'Information';

export const Warning = Template.bind({});
Warning.args = { text: 'Warning', type: 'warning', hasIcon: true };
Warning.storyName = 'Warning';

export const Erreur = Template.bind({});
Erreur.args = { text: 'Erreurs', type: 'error', hasIcon: true };
Erreur.storyName = 'Erreur';

export const Petit = Template.bind({});
Petit.args = { text: 'Petit', type: 'success', isSmall: true };
Petit.storyName = 'Petit';

export const Icon = Template.bind({});
Icon.args = { text: 'Icon', type: 'success', hasIcon: false };
Icon.storyName = 'Sans icone';
