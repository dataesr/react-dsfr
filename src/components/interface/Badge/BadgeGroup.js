import React, { Children } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import dataAttributes from '../../../utils/data-attributes';
import typeValidation from '../../../utils/type-validation';

/**
 *
 * @visibleName Badge
 */
const BadgeGroup = ({
  children,
  className,
  ...remainingProps
}) => {
  const content = Children.toArray(children)
  // eslint-disable-next-line react/no-array-index-key
    .map((child, index) => <li key={index}>{child}</li>);
  const _className = classNames('fr-badges-group', {}, className);
  return (
    <ul
      className={_className}
      {...dataAttributes.getAll(remainingProps)}
    >
      {content}
    </ul>
  );
};

BadgeGroup.defaultProps = {
  children: '',
  __TYPE: 'BadgeGroup',
  className: '',
};

BadgeGroup.propTypes = {
  __TYPE: typeValidation('BadgeGroup'),
  className: PropTypes.string,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
};

export default BadgeGroup;
