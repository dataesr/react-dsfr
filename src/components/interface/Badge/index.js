import Badge from './Badge';
import BadgeGroup from './BadgeGroup';

export { Badge, BadgeGroup };
