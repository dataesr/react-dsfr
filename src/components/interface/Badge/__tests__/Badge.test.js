import { render, screen } from '@testing-library/react';
import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import Badge from '../Badge';

Enzyme.configure({ adapter: new Adapter() });

describe('<Badge />', () => {
  it('should render Badge properly', () => {
    render(
      <Badge
        className="custom-class"
        data-testid="badge"
        text="text"
        type="success"
        icon={false}
        small
      />,
    );
    const badge = screen.getByTestId('badge');
    expect(badge).toMatchSnapshot();
  });
});
