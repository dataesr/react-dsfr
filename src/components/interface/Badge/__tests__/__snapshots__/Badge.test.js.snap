// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Badge /> should render Badge properly 1`] = `
<p
  class="fr-badge fr-badge--success fr-badge--no-icon custom-class"
  data-testid="badge"
>
  text
</p>
`;
